### PR TITLE
Better handled Spectrum1D images across classes

### LINKED ISSUE #129: Background errors out when given an NDData-typed image
This happens because `NDData` images don't have a "first-level" `shape` attribute like numpy `array` and `CCDData` objects do. If `image` is an `NDData` object, you have to go down to `image.data.shape` for the equivalent information.

Is this something we should handle here, or should we try to petition Astropy to expose `shape` as directly accessible on `NDData` objects?

### LINKED ISSUE #130: Extend Trace/Background/Extract to support specutils.Spectrum1D objects as inputs
this would be useful, especially for tracking units and so that wavelength solutions can be propagated to the returned objects.

### PR DESCRIPTION
**Pull request by ojustino**
I made changes to `Background` and the `Trace` classes in particular.

I created ~an `__init__()`~ a private method for `Trace` and put the check for `Spectrum1D` there so it can be propagated out to the children classes through `super()`. I'm not sure if I should be using `__post_init__()` instead.

I didn't realize that `Spectrum1D` is a child class of `NDData`, so most of the checks for the latter in `Background` and the extraction classes already captured them. I added code in `Background` to fix #129 and made the docstrings on image typing consistent across all classes.

I hope this fixes #130, but I wasn't able to reproduce the strange behavior with units reported there.

### CODE REVIEW

**Comment by ojustino** [comment]
As far as I can tell, this approach works with images that are bare numpy arrays, along with Spectrum1D, NDData, and CCDData objects. It doesn't work with images that are Quantity objects due to because various places in our code compare or do arithmetic with the image and another unit-less array.

Should the Quantity scenario be fixed or should we mandate that users provide an NDData-inheriting object if they want their image to have units?

**Review by kecnry — COMMENTED** [review]
Just a few questions on the implementation and then I will try to robustly test and see if I can reproduce any of the previous units issues.

All of the extra logic on how to handle the unit makes me wonder if we should _only_ support passing the unit in _with_ the image (as either a NDData or Spectrum1D or quantity array) and removing the `unit` input parameter entirely.  Any thoughts on that?

File: specreduce/extract.py
Comment (ojustino) [reply specreduce/extract.py:183]: I don't think this works because numpy arrays have a ``data` attribute that returns a buffer object instead of what we'd like.

Comment (kecnry) [reply specreduce/extract.py:183]: ah, good point.  I suppose we could check the type of that, but then maybe the logic is getting just as, if not more, complicated than what you already have here.

File: specreduce/tracing.py
Comment (ojustino) [reply specreduce/tracing.py:44]: This drops the unit. Are you thinking of saving it as a class attribute or something else?

File: specreduce/background.py
Comment (ojustino) [reply specreduce/background.py:91]: (This also drops the unit.)

File: specreduce/tracing.py
Comment (kecnry) [reply specreduce/tracing.py:44]: Well this is the main reason why I was wondering if we should keep the quantity/NDData/spectrum1D internally as opposed to the separate unit parameter, that way we can keep the unit internally and avoid all the parsing logic.

Comment (ojustino) [reply specreduce/tracing.py:44]: That will be more involved, but it would probably also fix the Quantity issue I mentioned in my second standalone comment in the PR thread. I'll give it a try and see how it comes out.

Comment (ojustino) [reply specreduce/tracing.py:44]: I think skipping the parsing logic pushes the problem farther down the road. `self.image` is just the flux when you have a CCDData/NDData/numpy array image but is an object containing the flux, spectral axis, and uncertainty when `self.image` is type Spectrum1D. The only way to put them all on the same footing is to take `image.data` of the NDData-inheriting types (which include Spectrum1D), which produces a unit-less array.

If you don't pre-process, then you have to check type any time you want to do an operation (e.g., if it's an array, I can take `image.mean()`, but if it's Spectrum1D, I have to do `image.flux.mean()` or `image.data.mean()` instead, etc.).

Comment (kecnry) [reply specreduce/tracing.py:44]: ok, I agree then, some sort of parsing and coercing into a consistent internal type is probably the best way to go.  The questions that remain then are:
* what is the most convenient type to use internally?  If we ever are going to act on uncertainties, if provided, then maybe converting any input into `Spectrum1D` objects and storing that internally makes sense?
* do we always output the same type?  Or do we attempt to convert back to the type the user provided?  This might be out of scope for this PR, but at least worth planning ahead.
* does it make sense to still have the `unit` input as a dedicated parameter or require the user to pass an object that contains units if they don't want default units?  I think I prefer removing (probably with deprecation) the `unit` input parameter to avoid user-confusion and internal logic needed to handle the case where the parameter and a passed `Spectrum1D` or quantity object have conflicting units.

Comment (ojustino) [reply specreduce/tracing.py:44]: To this point, the practice has been to give back `self.image` the same way the user provided it as its class' `image` argument. Then, inside the class, we use `self.image` to create a local variable that, if it's not already one, converts the image to a numpy array (either normal or masked) for internal operations.

If we accept Spectrum1D images, we probably need to start coercing all images to a consistent type before assigning them to `self.image`, since, as I mentioned above, calling a Spectrum1D object returns more than just its flux. An upside of this approach is that it would also allow us to use `self.image` for internal operations instead of a converted local variable since we'd now be sure of its type. 

CCDData is my pick as the best type for that purpose. Working internally with unit-less numpy arrays would be simpler, but if a user provides an image with units, they should be able to see those units when they call `self.image`. Unlike Quantity arrays, you could also get the uncertainty if you dig deeper (`self.image.uncertainty`), which is a plus.

`HorneExtract` is the only class that takes a `unit` argument, and it was requested so that authors had the ability to pass a string unit with a regular numpy array and get back a spectrum with Astropy units. The `unit` argument only matters if `image` is a numpy array; like `variance` and `mask`, it is skipped otherwise.

Comment (kecnry) [reply specreduce/tracing.py:44]: CCDData sounds like a reasonable pick to me.  The other option that might be worth considering is Spectrum1D since I think some methods currently always return a Spectrum1D and these packages are intended to coordinate well with each other.  If that comes with more overhead though, then using CCDData internally and mapping to Spectrum1D when necessary for output works for me.

> HorneExtract is the only class that takes a unit argument, and it was requested so that authors had the ability to pass a string unit with a regular numpy array and get back a spectrum with Astropy units. The unit argument only matters if image is a numpy array; like variance and mask, it is skipped otherwise.

I would still be in favor of deprecating this, but it doesn't have to happen in this PR.  If we ever did get rid of it, the user would still be able to use `numpy_array * u.Unit(unit_string)`, and I think it would considerably clean up internal logic once we were able to get rid of it fully.  Alternatively, adding matching support across all other classes would either require duplicate code logic or refactoring with some base classes or decorators that handle that in general.

**Comment by ojustino** [comment]
I added a parser to all trace, extraction, and background classes so they can handle input `Spectrum1D` and `Quantity` type images in addition to their existing support for `CCDData`, `NDData`, and numpy arrays. Most parsers are the same, though `HorneExtract`'s is more stringent. I could try to iron out the differences if we would rather have all classes inherit the parser from `SpecreduceOperation`.

The `image` attribute is now always of type `Spectrum1D`. I went this route because `Spectrum1D` has spectral axis information that would be lost with any other type. Inputs that lack spectral axes are just given an arange of values in pixel units. 

Some questions:

1. Before, `obj.image` was stored in our classes the way it was introduced as an argument. Is it an annoying user experience to have it converted to `Spectrum1D`?
2. As a consequence of the `image` attribute's type change, `obj.image` is now a conglomerate of information instead of just the flux. This makes getting to the flux for plotting, arithmetic, etc. a bit more cumbersome for those who don't know `Spectrum1D` (e.g., from `plt.imshow(obj.image)` before to `plt.imshow(obj.image.data)` now). Are we OK with that?
3. Would this be the time update `_to_spectrum1d_pixels()` so it takes the predefined spectral axis from the `image` attribute instead of creating a spectral axis of its own?
4. Should flux units for images that lack them continue to be counts or switch to blank units (`u.Unit()`)? Most classes currently default to counts, though `HorneExtract` defaults to blank units.

I will push some new tests of image input typing soon. I'll wait on adjusting the documentation until it's clear that the new approach is acceptable.

**Comment by ojustino** [comment]
Looks like I have some failures to address before I introduce the new tests. Thoughts on the new approach to typing are still welcome.

**Review by kecnry — COMMENTED** [review]
> The image attribute is now always of type Spectrum1D. I went this route because Spectrum1D has spectral axis information that would be lost with any other type. Inputs that lack spectral axes are just given an arange of values in pixel units.

I think this is a good idea and makes sense with the goal to integrate more tightly with specutils... but share your concerns below.

> Before, obj.image was stored in our classes the way it was introduced as an argument. Is it an annoying user experience to have it converted to Spectrum1D?
> As a consequence of the image attribute's type change, obj.image is now a conglomerate of information instead of just the flux. This makes getting to the flux for plotting, arithmetic, etc. a bit more cumbersome for those who don't know Spectrum1D (e.g., from plt.imshow(obj.image) before to plt.imshow(obj.image.data) now). Are we OK with that?

An alternative, I suppose, might be to store the user-provided object and then do the conversion (`_parse_image`) as part of a (cached) property.  That way both formats are accessible to the user, but internally we can rely on a consistent type when calling the property.

File: specreduce/extract.py
Comment (ojustino) [reply specreduce/extract.py:170]: What do you think of adding it as a class method in `SpecreduceOperation`? All the classes that need it inherit from there. `HorneExtract` needs a more stringent parser, but I could override the inherited `_parse_image()` by redefining it in `HorneExtract`.

Comment (kecnry) [reply specreduce/extract.py:170]: > What do you think of adding it as a class method in SpecreduceOperation?

I think that's a good idea if everything that inherits from `SpecreduceOperation` would make use of it... otherwise you could have a base class between that inherits from `SpecreduceOperation` and adds image functionality.  And I think even if you want to switch to the property idea, that could easily be migrated within that base class.

**Comment by ojustino** [comment]
I had more to do since my last messages than I thought:

* I created the `_ImageParser` to serve as an ancestor class that houses the code for converting user-provided images to consistently formatted `Spectrum1D` objects in all the operations. I'm open to other design suggestions on inheritance if this isn't ideal.
* I made a lot of modifications to `Background` and existing tests so they play better with the switch to `Spectrum1D` images.
* I wrote new tests that specifically cover the image parsing process. Images with the same flux, uncertainty, etc. but in different formats should be exactly the same once they've gone through `_parse_image()`. Image objects that lack certain attributes (e.g., `CCDData` objects with `mask == None` or that numpy arrays always lack an `uncertainty`) should see them added to the parsed image object in a consistent manner.
* I added a requirement of `specutils>=1.7` to avoid having NaNs appear in the spectral axis of the result when you use `Spectrum1D.subtract()` on an image that's been through `_parse_image()`, such as in `Background.sub_image()`. We also need `asdf_astropy>=0.2.0` for that, but it looks like the `specutils` requirement gives us that for free.

The notebooks will need edits, but I would like to see #128 merged first to avoid merge conflicts.

Questions:
1. Should `Background.bkg_array` still be a numpy array given the rest of the transition to  `Spectrum1D`?
2. Should `_parse_image()` be called in `__call__` for the *Extract classes? When it happens there, you can never actually retrieve the parsed version of the original image because `hrn.image` is the user's unparsed input to `hrn = HorneExtract(image, trace)` and the result of `hrn()` is the extracted image as a `Spectrum1D` object. Does that matter?
3. Should flux units for images that lack them be in counts or blank units (`u.Unit()`)? Most classes currently default to counts, though `HorneExtract` defaults to blank units.

External dependency:

- We need functionality from astropy/specutils#988 for subtraction of a `Background` object to work on `Spectrum1D` objects. This is handled in `Background.__rsub__`, but `Spectrum1D.__sub__` has to give it the chance to work by passing the operation. I have commented out a related test in the meantime.

File: specreduce/extract.py
Comment (ojustino) [reply specreduce/extract.py]: I forgot that only the extraction operations inherit from `SpecreduceOperation`, so I tried putting `_parse_image` in an mixin class used as an ancestor for any class that needs its functionality instead. More details in the summary I posted below.

File: specreduce/background.py
Comment (ojustino) [reply specreduce/background.py]: Agreed. I tried your idea of storing `image` as a `cached_property`, but I couldn't figure out a way to do it without either a) changing the name of the `image` argument, which I think would be unintuitive, or b) including a dummy argument with `InitVar` to be converted to `image` later on. The dummy argument either shows up in the automatic docstring (confusing to users) or isn't passed to the `__init__`.

Maybe there's some potential crossover with #147 on this topic?

Also, it looks like `cached_property` requires Python ≥3.8. We currently allow 3.7, but the `specutils` requirements I mention in my summary below would give us an implicit requirement of 3.8, too.

File: specreduce/background.py
Comment (kecnry) [inline specreduce/background.py:286]: this line seems to be dropping the units on the spectral axis (if I pass `image` as a `Spectrum1D` with wavelength units, `sub_image()` is returning in pixels and have confirmed that both `image` and `self.bkg_image(image)` still have the correct wavelength units).  Removing `compare_wcs=None` fixes the problem, so we might need something like the following since `compare_wcs=None` is needed to support the pixels-case (feel free to refactor however makes sense to you):

```
from astropy import units as u
bkg_image = self.bkg_image(image)
kwargs = {'compare_wcs': None} if bkg_image.spectral_axis.unit == u.pix else {}
return image.subtract(self.bkg_image(image), **kwargs)
```

Also see https://github.com/astropy/specutils/issues/989 which was found while debugging this but can probably be ignored here with the assumption that specutils will properly handle the spectral axes comparison in the future.

Comment (ojustino) [reply specreduce/background.py]: That makes sense. I got this working in the latest commit and also added tests that would have caught the situation you encountered here. I agree with deferring a spectral axis compatibility check to specutils.

Comment (kecnry) [reply specreduce/background.py]: awesome, thanks!  Can confirm that it now works as expected for the wavelength-units case

**Review by kecnry — APPROVED** [review]
This looks good to me and works in the cases I've tested.  This probably counts as an API break since the output types have changed, but in the past we made similar breaking changes in minor releases since we're still considered in active development and not adhering to semver rules... but just something to keep in mind.

**Comment by kecnry** [comment]
> Should `Background.bkg_array` still be a numpy array given the rest of the transition to `Spectrum1D`?

I think so since it's literally called array, but I also am not opposed to removing it (maybe with a deprecation period) and replacing it with something named more appropriately to hold a `Spectrum1D` object (either in this PR or down the road).

> Should `_parse_image()` be called in `__call__` for the *Extract classes? When it happens there, you can never actually retrieve the parsed version of the original image because `hrn.image` is the user's unparsed input to `hrn = HorneExtract(image, trace)` and the result of `hrn()` is the extracted image as a Spectrum1D object. Does that matter?

I'm a bit torn on this one.  I proposed the cached property idea, but that seemed to not play nicely with dataclasses.  Perhaps if we decide to abandon dataclasses (see https://github.com/astropy/specreduce/issues/126#issuecomment-1307269176), then that is a direction we could take in the future (in which case we could have `hrn.orig_image` be the original input and then `hrn.image` be a cached property where the parsing takes place and is made available to `__call__`).

> Should flux units for images that lack them be in counts or blank units (u.Unit())? Most classes currently default to counts, though HorneExtract defaults to blank units.

I think this is out-of-scope for this PR, but maybe worth opening a follow-up issue so that we can discuss and decide.  I agree that we should aim to be self-consistent.

**Comment by tepickering** [comment]
> Just a few questions on the implementation and then I will try to robustly test and see if I can reproduce any of the previous units issues.
> 
> All of the extra logic on how to handle the unit makes me wonder if we should _only_ support passing the unit in _with_ the image (as either a NDData or Spectrum1D or quantity array) and removing the `unit` input parameter entirely. Any thoughts on that?

my preference from the start was that units should be attached to the data rather than passed as an input parameter. it's just a lot easier and cleaner that way.

File: specreduce/background.py
Comment (tepickering) [reply specreduce/background.py]: if we haven't yet, we should officially update our python requirement to ≥3.8. especially since it's implicitly required by dependencies.

File: CHANGES.rst
Comment (tepickering) [inline CHANGES.rst:12]: i assume this will be filled in a bit more before merging.

File: setup.cfg
Comment (tepickering) [inline setup.cfg:21]: i'd suggest changing `python_requires` to 3.8 to be consistent with this.

File: specreduce/core.py
Comment (tepickering) [inline specreduce/core.py:73]: i think a default of `DN` is appropriate. it is a generic digital number that's appropriate to use if the provenance isn't known.

File: specreduce/background.py
Comment (tepickering) [inline specreduce/background.py:158]: this is confusing since `crossdisp_axis` can still be provided as an input...

File: specreduce/background.py
Comment (tepickering) [inline specreduce/background.py:210]: is this parameter being used or is the assumption hard-coded as implied in the previous docstring?

File: specreduce/extract.py
Comment (tepickering) [inline specreduce/extract.py:384]: i have a light preference for using `DN` as a default, but blank/dimensionless is ok. whatever choice is made, it should be consistently applied.

**Review by tepickering — CHANGES_REQUESTED** [review]
impressive amount of work here! i advocated early on for the adoption of `NDData` as the standard data structure to use internally. this PR takes that even further by adopting a more appropriate subclass which i 💯 agree with. i appreciate the clean-up by putting the image parsing all in one place. i have a few minor comments to address and note that there are conflicts that need to be resolved before merging. coverage looks good and the tests pass so once that's taken care of, it's gtg to merge. thanks again for all the hard work on this!

File: CHANGES.rst
Comment (ojustino) [reply CHANGES.rst]: I can include that operations' `image` attributes will now always be `Spectrum1D` objects. Was there anything else you were thinking should be here? I thought many of the changes were too behind the scenes to make the changelog.

File: specreduce/extract.py
Comment (ojustino) [reply specreduce/extract.py]: I changed to DN here to match the rest of the operations.

File: specreduce/background.py
Comment (ojustino) [reply specreduce/background.py:158]: I agree, but the extraction operations function the same way. I think it's a remnant of the desire to future-proof them for a time when we allow for more flexibility in axis arrangement. Would you agree that a solution to this is outside this PR's scope?

File: specreduce/background.py
Comment (ojustino) [reply specreduce/background.py:210]: `cls` should be of type `Background`, so it follows the docstring's implication once the new instance is created at the end of the method.

**Comment by ojustino** [comment]
@tepickering, I believe I addressed everything you brought up that's in scope here. The failing test is due to an upstream issue with Astropy that looks like it will be resolved soon.

I do feel like this is ready for a merge... but also feel it's bad juju to merge one's own pull request with a failing test and changes requested, so I'll defer to others' judgment or until after the holiday.

**Comment by ojustino** [comment]
OK, I will merge now to keep things moving.

**Pull request merged**
